### PR TITLE
Porting to proxy

### PR DESCRIPTION
**Pull request by dfquaresma**

### CODE REVIEW

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java [hunk 1]
```diff
@@ -29,19 +27,24 @@
...
-		if (shedResponse.shouldShed) {
-			response.setStatus(HttpStatus.SERVICE_UNAVAILABLE.value());
-			response.setContentLength(0);
+		
+		switch (request.getHeader(GCI_HEADERS_NAME)) {
+		case "":
+			return true;
+
+		case "ch":
+			response.getWriter().write((int) gci.getHeapUsageSinceLastGC());
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java]: This should be "young | old"

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java [hunk 2]
```diff
@@ -29,19 +27,24 @@
...
+			response.getWriter().flush();
+			response.getWriter().close();
+			response.setStatus(HttpStatus.OK.value());
+			return false;
+			
+		default:
+			response.setStatus(HttpStatus.OK.value());
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java]: Response should be set after the gc.

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java [hunk 3]
```diff
@@ -20,7 +19,6 @@
 @Configuration
 public class SpringGcInterceptor extends HandlerInterceptorAdapter {
 
-	private static final String SHED_RESPONSE_OBJ_NAME = "sr";
 	private static final String GCI_HEADERS_NAME = "GCI";
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java]: "gci"

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java [hunk 4]
```diff
@@ -29,19 +27,24 @@
...
-		ShedResponse shedResponse = gci.before(request.getHeader(GCI_HEADERS_NAME));
-		if (shedResponse.shouldShed) {
-			response.setStatus(HttpStatus.SERVICE_UNAVAILABLE.value());
-			response.setContentLength(0);
+		
+		switch (request.getHeader(GCI_HEADERS_NAME)) {
+		case "":
+			return true;
+
+		case "ch":
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java]: Make "ch" a constant.

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java [hunk 5]
```diff
@@ -29,19 +27,24 @@
 	@Override
 	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
 			throws Exception {
-		ShedResponse shedResponse = gci.before(request.getHeader(GCI_HEADERS_NAME));
-		if (shedResponse.shouldShed) {
-			response.setStatus(HttpStatus.SERVICE_UNAVAILABLE.value());
-			response.setContentLength(0);
+		
+		switch (request.getHeader(GCI_HEADERS_NAME)) {
+		case "":
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java]: Please check the APIs (https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getHeader(java.lang.String))

This method returns null if the header is not present. It is different from the ruby case :)

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java [hunk 6]
```diff
@@ -29,19 +27,24 @@
...
+			return false;
+			
+		default:
+			response.setStatus(HttpStatus.OK.value());
+			gci.collect();
+			return false;
 		}
-		request.getSession().setAttribute(SHED_RESPONSE_OBJ_NAME, shedResponse);
-		return !shedResponse.shouldShed;
+
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptor.java]: Remove this blank line.

File: core/src/main/java/com/gcinterceptor/core/RuntimeEnvironment.java [hunk 7]
```diff
@@ -4,79 +4,50 @@
...
 	public long getYoungHeapUsageSinceLastGC() {
-		youngLastUsed = getYoungHeapUsage() - youngLastAlloc;
-		return youngLastUsed;
+		return getYoungHeapUsage() - youngLastAlloc;
 	}
 
 	public long getTenuredHeapUsageSinceLastGC() {
-		tenuredLastUsed = getTenuredHeapUsage() - tenuredLastAlloc;
-		return tenuredLastUsed;
+		return getTenuredHeapUsage() - tenuredLastAlloc;
```
Comment (danielfireman) [inline core/src/main/java/com/gcinterceptor/core/RuntimeEnvironment.java]: Why is this necessary? This is not what the proxy wants.

File: spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptorConfiguration.java [hunk 8]
```diff
@@ -21,7 +21,7 @@ public SpringGcInterceptor springGcInterceptor() {
     }
 
     @Bean
```
Comment (danielfireman) [inline spring/src/main/java/com/gcinterceptor/spring/SpringGcInterceptorConfiguration.java]: This returned could be a singleton. We don't want to loop over the memory pools for every request.

**Comment by dfquaresma** [comment]
@danielfireman, could you please review one more time now?
